Update assimp version

NolaDonato opened the pull request:
Replaces pull request #1027 which I seem to have made a mess of. This is a merged, squashed commit of that.

- Adds libraries for the latest version of assimp.
- includes two bug fixes to assimp involving OBJ import.
- Bug fix in asset loader for UV indices sometimes starting at 1 instead of 0
- Eliminated redundant version of assimp used for loadMesh (reimplemented loadMesh in terms of loadModel)
- Fixed a framework bug with TGA import.

Passes all the unit tests, has been tested with lots of demos. Needs more testing.

GearVRF DCO signed off by: Nola Donato nola.donato@samsung.com

liaxim commented:
gvr-3dcursor always crashes on launch for me. My device is Note 5, GLES: ARM, Mali-T760, OpenGL ES 3.1 v1.r7p0-03rel0.b596bd02e7d0169c10574b57180c8b57.

```
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410] JNI DETECTED ERROR IN APPLICATION: JNI FindClass called with pending exception java.lang.NullPointerException: Attempt to invoke virtual method 'void java.io.InputStream.close()' on a null object reference
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at void org.gearvrf.GVRAndroidResource.closeStream() (GVRAndroidResource.java:312)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at byte[] org.gearvrf.GVRAssetLoader$ResourceVolumeIO.read(java.lang.String) (GVRAssetLoader.java:892)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at byte[] org.gearvrf.GVRAssetLoader$CachedVolumeIO.read(java.lang.String) (GVRAssetLoader.java:918)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at org.gearvrf.jassimp.AiScene org.gearvrf.jassimp.Jassimp.aiImportFileEx(java.lang.String, long, org.gearvrf.jassimp.JassimpFileIO) (Jassimp.java:-2)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at org.gearvrf.jassimp.AiScene org.gearvrf.jassimp.Jassimp.importFileEx(java.lang.String, java.util.Set, org.gearvrf.jassimp.JassimpFileIO) (Jassimp.java:123)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at org.gearvrf.GVRSceneObject org.gearvrf.GVRAssetLoader.loadJassimpModel(org.gearvrf.GVRAssetLoader$AssetRequest, org.gearvrf.GVRSceneObject, java.util.EnumSet, boolean, org.gearvrf.GVRScene) (GVRAssetLoader.java:1339)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at org.gearvrf.scene_objects.GVRModelSceneObject org.gearvrf.GVRAssetLoader.loadModel(org.gearvrf.GVRAndroidResource, java.util.EnumSet, boolean, org.gearvrf.GVRScene) (GVRAssetLoader.java:1253)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at org.gearvrf.GVRMesh org.gearvrf.GVRContext.loadMesh(org.gearvrf.GVRAndroidResource, java.util.EnumSet) (GVRContext.java:301)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at org.gearvrf.GVRMesh org.gearvrf.GVRContext.loadMesh(org.gearvrf.GVRAndroidResource) (GVRContext.java:256)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at org.gearvrf.GVRMesh org.gearvrf.asynchronous.AsyncMesh$AsyncLoadMesh.loadResource() (AsyncMesh.java:101)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at java.lang.Object org.gearvrf.asynchronous.AsyncMesh$AsyncLoadMesh.loadResource() (AsyncMesh.java:85)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at void org.gearvrf.asynchronous.Throttler$AsyncLoader.run() (Throttler.java:184)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at void org.gearvrf.asynchronous.Throttler$PendingRequests$PendingRequest.run() (Throttler.java:507)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at void org.gearvrf.utility.Threads$ThreadLimiter$ThreadManager.run() (Threads.java:521)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at void org.gearvrf.utility.Threads$1.run() (Threads.java:108)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at java.lang.Object java.util.concurrent.Executors$RunnableAdapter.call() (Executors.java:423)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at void java.util.concurrent.FutureTask.run() (FutureTask.java:237)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at void java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) (ThreadPoolExecutor.java:1113)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at void java.util.concurrent.ThreadPoolExecutor$Worker.run() (ThreadPoolExecutor.java:588)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]   at void java.lang.Thread.run() (Thread.java:818)
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410] 
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]     in call to FindClass
02-22 14:46:24.782  6970  7070 F art     : art/runtime/java_vm_ext.cc:410]     from org.gearvrf.jassimp.AiScene org.gearvrf.jassimp.Jassimp.aiImportFileEx(java.lang.String, long, org.gearvrf.jassimp.JassimpFileIO)
```

NolaDonato commented:
The 3D cursor example simultaneously loads multiple meshes from a ZIP file. I didn't even know we could do this :-) The old importer supported this with a separate version of assimp distinct from the one the asset loader uses with a completely separate code path to call it. When I changed loadMesh to use the asset loader, this case got dropped on the floor because the asset loader input path doesn't support streams. I added code to allow a GVRResourceVolume to come from an input stream and now the sample works. We may find other stuff like this lurking in other samples.

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/GVRActivity.java:
```diff
@@ -139,6 +139,19 @@ protected void onCreate(Bundle savedInstanceState) {
         setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
 
         mRenderableViewGroup = (ViewGroup) findViewById(android.R.id.content).getRootView();
+        mDockEventReceiver = new DockEventReceiver(this,
```
All the changes to GVRActivity should be reverted. This resurrects an older version of it.

liaxim commented:
The demos check out, almost. Have few more things to try.

There is an occasional jni crash with gvr-3dcursor. Sometimes it happens frequently sometimes it doesn't at all. Regardless the jni crash I can see quite often with this sample that GVRAndroidResource.closeStream is trying to close an OPEN stream but the stream is null hence NPE. The NPE might not be handled before a JNI call is made and that leads to the random crashes.

It seems to me related to this change. I don't see it with master. Just add a log statement in closeStream when a stream is OPEN but the stream member is null.

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/GVRAndroidResource.java:
```diff
@@ -308,7 +327,7 @@ public synchronized final InputStream getStream() throws IOException{
     public synchronized final void closeStream() {
         try {
-            if (streamState == StreamStates.OPEN) {
+            if ((stream != null) && (streamState == StreamStates.OPEN) {
```
What is this case where the state is OPEN but there is no stream? Is it a valid state to be in?

liaxim commented:
Says "This branch cannot be rebased due to conflicts"

NolaDonato commented:
I rebased again, now there are no conflicts.

liaxim commented:
Still "This branch cannot be rebased due to conflicts"

NolaDonato commented:
Tomorrow we should get together and investigate why yours always says it needs rebase and mine says there are no conflicts.

liaxim commented:
It can be merged but cannot be rebased since there are problems when rebasing the individual commits from this pull request. By default the rebase option is selected for me and it is grayed out. Anyway, we can proceed with the testing and merge instead.

liaxim commented:
Ok, all the additional things I wanted to try work. But we will have to properly rebase this otherwise there are things that will be overwritten with older versions (GVRActivity for example). Let me know which commit ids out of the 8 commits you truly want in this pull request and I will rebase it.

liaxim commented:
Merge squashed commits d43fdb5cdd6927caa18cbc5390454bd7fc5906fd, e11710ce8326050f9792927197d258aecaf42063, 687f3b669650f9c01f853efbe730987ba84043c1, bfb69d325fb862ba1c8d2c1a3d95de9d256bf67b, 105cf9c2e5a9f5b813a0bd20c690b6b58fc63311
Reverted the changes to GVRActivity and removed GearVRf/CMakeLists.txt

liaxim commented:
Looks good. Will merge Wednesday afternoon.
